R/baselinenowcast.R

```diff
@@ -152,3 +153,258 @@ baselinenowcast.reporting_triangle <- function(
+#'   unique combinations of reference dates and report dates. Default is `NULL`
+#'   which assumes that the data.frame being passed in represents a single
+#'   strata (only one nowcast will be produced). All columns that are not
+#'   part of the `strata_cols` will be removed.
+#' @param strata_sharing Vector of character strings. Indicates if and what
+#'   estimates should be shared for different nowcasting steps. Options are
+#'   `"none"` for no sharing (each `strata_cols` is fully independent),
+#'   `"delay"` for delay sharing and `"uncertainty"` for uncertainty sharing.
+#'   Both `"delay"` and `"uncertainty"` can be passed at the same time.
+#' @param ... Additional arguments passed to
+#'    \code{\link{estimate_uncertainty}}
+#'    and \code{\link{sample_nowcast}}.
+#' @inheritParams baselinenowcast
+#' @inheritParams as_reporting_triangle.data.frame
+#' @inheritParams estimate_uncertainty
+#' @inheritParams sample_nowcast
+#' @inheritParams allocate_reference_times
+#' @importFrom purrr imap list_rbind map_dfc set_names
+#' @importFrom checkmate assert_subset assert_character assert_names
+#' @family baselinenowcast_df
+#' @export
+#' @method baselinenowcast data.frame
+#' @returns Data.frame of class \code{\link{baselinenowcast_df}}
+#' @examples
+#' covid_data_to_nowcast <- germany_covid19_hosp[
+#'   germany_covid19_hosp$report_date <
+#'     max(germany_covid19_hosp$reference_date),
+#' ] # nolint
+#' nowcasts_df <- baselinenowcast(covid_data_to_nowcast,
+#'   max_delay = 40,
+#'   strata_cols = c("age_group", "location")
+#' )
+#' nowcasts_df
+baselinenowcast.data.frame <- function(
+    data,
+    scale_factor = 3,
+    prop_delay = 0.5,
+    output_type = c("samples", "point"),
+    draws = 1000,
+    uncertainty_model = fit_by_horizon,
+    uncertainty_sampler = sample_nb,
+    max_delay,
+    delays_unit = "days",
+    strata_cols = NULL,
+    strata_sharing = "none",
+    ...) {
+  output_type <- arg_match(output_type)
+  assert_names(colnames(data),
+    must.include = c("reference_date", "report_date", "count")
+  )
+  assert_character(delays_unit)
+  assert_character(strata_sharing)
+  assert_character(output_type)
+  assert_subset(strata_sharing,
+    choices = c("delay", "uncertainty", "none")
+  )
+  if (length(strata_sharing) == 2 && "none" %in% strata_sharing) {
+    cli_abort(
+      message = c("`strata_sharing` cannot be both 'none' and 'delay'/'uncertainty'") # nolint
+    )
+  }
+  # Filter to max delay
+  data$delay <- as.numeric(
+    difftime(
+      as.Date(data$report_date),
+      as.Date(data$reference_date),
+      units = delays_unit
+    )
+  )
+  data_clean <- data[data$delay <= max_delay, ]
+
+  .validate_strata_cols(
+    strata_cols,
+    data_clean
+  )
+  .validate_date_cols(data_clean, "reference_date")
+  .validate_date_cols(data_clean, "report_date")
+
+  # Split dataframe into a list of dataframes for each nowcast unit
+  if (length(strata_cols) != 0) {
```

Comment:
I think in an ideal world this would be an internal function that was tested.

Reply:
can make it a function and add tests